### ISSUE TITLE
How to not duplicate data without querying

### ISSUE THREAD

**Issue by lcasassa** [dgraph-io/dgraph#3517]
Hi All. Thanks for DGraph! Awesome documentation, awesome how easy it is to get it running. Looking forward in using it in production.

Im just starting with graphs and I'm lost on how to not create duplicates without doing queries. I should be able to it as I have unique values for each node. For example:

```
{
    "set": [
        {
            "uid": "0x1234",   # the client has a unique hash for this node
            "industry": {
                "industry.name": "Machinery",   # this is unique for industry node type
            }
            "company.name": "CompanyABC"
        },
        {
            "uid": "0x6789",
            "industry": { # same as the one above, don't want a duplicate
                "industry.name": "Machinery",
            }
            "company.name": "CompanyDEF"
        },
    ]
}
```

I understand I can use _:machinery to reference the node so it does not duplicate it but I can't use that for another request. For example after the fist mutations I will like to do this one:

```
{
    "set": [
        {
            "uid": "0x9abc",
            "industry": { # same as the one above, don't want a duplicate
                "industry.name": "Machinery",
            }
            "company.name": "CompanyJKL"
        },
        {
            "uid": "0x1234", # updating the node already created without changing the industry
            "company.name": "newCompanyABC"
        },
    ]
}

```
Should I use uid for the industry node? I could hash the name and get a unique uid on the client side.
I could use external id but I will have to query for the uid and would like to avoid querying and also there is a race condition. If two clients queries to find the uid with the external id and that uid was not created before, both clientes will try to create it making a duplication instead of an insert and an update.

What am I missing?

**Comment by MichaelJCompton** [comment]
Hi,

Looks like what you want is upserts.  That gives you a query, then create/update in a single transaction.  See these two sections in the docs

https://docs.dgraph.io/howto/#upserts 
https://docs.dgraph.io/query-language/#upsert-directive

and this for how you'd set the edge to an existing node in a JSON mutation
https://docs.dgraph.io/mutations/#edges-between-nodes

Note we are also adding a way to do this in a single operation (https://github.com/dgraph-io/dgraph/pull/3412) which might help your case once it's done.

Hope that helps.

**Comment by lcasassa** [comment]
Cool! Thanks. It worked but it is very slow (3 per second). I manage to look for 5 uids in one query and update the 5 nodes in separate queries. I can't have two clients updating as if I do so I get error on the transaction and have to retry making it slow. I have 20 TB of data in a ES. What is my best option to import lots of data when I don't know the uid in advance?

**Comment by lcasassa** [comment]
https://docs.dgraph.io/deploy/#live-loader
-x option will save the xid to a directory so it will use the same uid and not duplicate on a second run. Will give that a try :D

**Comment by MichelDiz** [comment]
The new Upsert transaction is available in the v1.1 RCs. And will be part of the final release. The way this new transaction works will fit perfectly the needs of this question. Also, the mentioned -x option that saves the mapped UIDs to a file is a very handy option to use if you use the same Blank Nodes.

I consider this issue to be resolved and as it is an old issue I will close. Any questions feel free to ask.

**Comment by lcasassa** [comment]
Update with dgraph v1.1.1. We are able to get the results of the query on a upsert:

```
upsert {
  query {
    q(func: eq(external_id, "JF234")) {
      u as uid
    }
  }

  mutation {
    set {
      uid(v) <external_id> "JF234" .
      uid(v) <name> "Apple" .
    }
  }
}
```

In this way we can get the uid. It will be in two different places in the response depending if it is creating a new node or if it found the external_id in a node.
